error compiling wasm func ... failed to compile: BUG: relative jump offset

**Describe the bug**
Ran a program with the most recent tagged version of `wazero`, which failed with this error:

```sh
error: error compiling wasm func[.github.com_zonedb_zonedb.init]: failed to compile: BUG: relative jump offset -2097240/4(=-524310) must be within -524288 and 524287
```

https://github.com/tetratelabs/wazero/blob/b01ba67fdcf32997d22a0f56b723a71c1c40cd75/internal/asm/arm64/impl.go#L864

Prior to upgrading `wazero`, the program ran correctly.

**To Reproduce**
I was compiling a program that imports https://github.com/zonedb/zonedb.

**Expected behavior**
For the program to compile and run.


**Environment (please complete the relevant information):**
- Go version: `gotip` as of 2023-06-15
- wazero Version: v1.2.1
- Host architecture: arm64
- Runtime mode: compiler

**Additional context**
Using `wasirun` from https://github.com/stealthrocket/wasi-go

interesting, the patch causes `TestCompiler_compileBrTable` to fail

```
--- FAIL: TestCompiler_compileBrTable (0.00s)
    --- FAIL: TestCompiler_compileBrTable/only_default_with_index_0 (0.00s)
        require.go:330: expected 1, but was 6
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:316
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:477
    --- FAIL: TestCompiler_compileBrTable/only_default_with_index_100 (0.00s)
        require.go:330: expected 1, but was 6
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:316
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:477
    --- FAIL: TestCompiler_compileBrTable/select_default_with_targets_and_good_index (0.00s)
        require.go:330: expected 1, but was 4
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:316
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:477
    --- FAIL: TestCompiler_compileBrTable/select_default_with_targets_and_huge_index (0.00s)
        require.go:330: expected 1, but was 4
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:316
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:477
```

--- a/internal/engine/compiler/impl_arm64.go
+++ b/internal/engine/compiler/impl_arm64.go
@@ -380,42 +380,16 @@ func (c *arm64Compiler) compileReturnFunction() error {
 }
 
 func (c *arm64Compiler) compileMaybeExitFromNativeCode(skipCondition asm.Instruction, status nativeCallStatusCode) {
-	if target := c.compiledTrapTargets[status]; target != nil {
-		// We've already compiled this.
-		// Invert the condition to jump into the appropriate target.
-		var trapCondition asm.Instruction
-		switch skipCondition {
-		case arm64.BCONDEQ:
-			trapCondition = arm64.BCONDNE
-		case arm64.BCONDNE:
-			trapCondition = arm64.BCONDEQ
-		case arm64.BCONDLO:
-			trapCondition = arm64.BCONDHS
-		case arm64.BCONDHS:
-			trapCondition = arm64.BCONDLO
-		case arm64.BCONDLS:
-			trapCondition = arm64.BCONDHI
-		case arm64.BCONDHI:
-			trapCondition = arm64.BCONDLS
-		case arm64.BCONDVS:
-			trapCondition = arm64.BCONDVC
-		case arm64.BCONDVC:
-			trapCondition = arm64.BCONDVS
-		default:
-			panic("BUG: couldn't invert condition")
-		}
-		c.assembler.CompileJump(trapCondition).AssignJumpTarget(target)
-	} else {
-		skip := c.assembler.CompileJump(skipCondition)
-		c.compileExitFromNativeCode(status)
-		c.assembler.SetJumpTargetOnNext(skip)
-	}
+	skip := c.assembler.CompileJump(skipCondition)
+	c.compileExitFromNativeCode(status)
+	c.assembler.SetJumpTargetOnNext(skip)
 }
 
 // compileExitFromNativeCode adds instructions to give the control back to ce.exec with the given status code.
 func (c *arm64Compiler) compileExitFromNativeCode(status nativeCallStatusCode) {
 	if target := c.compiledTrapTargets[status]; target != nil {
 		c.assembler.CompileJump(arm64.B).AssignJumpTarget(target)
+		return
 	}
 
 	switch status {

--- a/internal/engine/compiler/impl_arm64_test.go
+++ b/internal/engine/compiler/impl_arm64_test.go
@@ -255,8 +255,11 @@ func TestArm64Compiler_LargeTrapOffsets(t *testing.T) {
 
 	// Repeat enough times that jump labels are not within (-524288, 524287).
 	// Relative offset -2097164/4(=-524291).
-	// 52429 is empirically the value that starts triggering the bug on arm64.
-	for i := 0; i < 52429; i++ {
+	// At the time of writing, 52429 is empirically the value that starts
+	// triggering the bug on arm64. We impose an arbitrarily higher value
+	// to account for possible future improvement to the number of instructions
+	// we emit.
+	for i := 0; i < 80_000; i++ {
 		err = compiler.compileConstI32(five)
 		require.NoError(t, err)
 

--- a/internal/engine/compiler/impl_arm64_test.go
+++ b/internal/engine/compiler/impl_arm64_test.go
@@ -236,6 +236,49 @@ func TestArm64Compiler_getSavedTemporaryLocationStack(t *testing.T) {
 	})
 }
 
+// https://github.com/tetratelabs/wazero/issues/1522
+func TestArm64Compiler_LargeTrapOffsets(t *testing.T) {
+	env := newCompilerEnvironment()
+	compiler := env.requireNewCompiler(t, &wasm.FunctionType{}, newCompiler, &wazeroir.CompilationResult{
+		Types: []wasm.FunctionType{{}},
+	})
+	err := compiler.compilePreamble()
+	require.NoError(t, err)
+
+	one := operationPtr(wazeroir.NewOperationConstI32(uint32(1)))
+	five := operationPtr(wazeroir.NewOperationConstI32(uint32(5)))
+	div := operationPtr(wazeroir.NewOperationDiv(wazeroir.SignedTypeInt32))
+
+	// Place the offset value.
+	err = compiler.compileConstI32(one)
+	require.NoError(t, err)
+
+	// Repeat enough times that jump labels are not within (-524288, 524287).
+	// Relative offset -2097164/4(=-524291).
+	// 52429 is empirically the value that starts triggering the bug on arm64.
+	for i := 0; i < 52429; i++ {
+		err = compiler.compileConstI32(five)
+		require.NoError(t, err)
+
+		err = compiler.compileDiv(div)
+		require.NoError(t, err)
+	}
+
+	err = compiler.compileReturnFunction()
+	require.NoError(t, err)
+
+	code := asm.CodeSegment{}
+	defer func() { require.NoError(t, code.Unmap()) }()
+
+	// Generate the code under test and run.
+	_, err = compiler.compile(code.NextCodeSection())
+	require.NoError(t, err)
+
+	env.exec(code.Bytes())
+
+	require.Equal(t, nativeCallStatusCodeReturned.String(), env.compilerStatus().String())
+}
+
 // compile implements compilerImpl.setStackPointerCeil for the amd64 architecture.
 func (c *arm64Compiler) setStackPointerCeil(v uint64) {
 	c.stackPointerCeil = v

